installDeps.sh: use "npm ci" instead of "npm install"?

In #3750 a user implicitly suggested we could use `npm ci` to install dependencies. At the moment, `bin/installDeps.sh` uses `npm install`:

https://github.com/ether/etherpad-lite/blob/1ad6e9c2880ed074ba7f6185bdb399d84b298310/bin/installDeps.sh#L103



Reading from the [`npm ci` docs](https://docs.npmjs.com/cli/ci.html):

> it’s meant to be used in [...] any situation where you want to make sure you’re doing a clean install of your dependencies.
> It can be **significantly faster** than a regular npm install by skipping certain user-oriented features.
> It is also more strict than a regular install, which **can help catch errors or inconsistencies caused by the incrementally-installed local environments** of most npm users.

## Possible advantages
1. **increased reproducibility** of the installations (both direct and via Docker);
2. once installed, **checking dependencies would be very fast**. This would solve the underlying issue below #3711.

   ~~Afterwards, we can still commit [`fastRun.sh`](https://github.com/ether/etherpad-lite/issues/3711#issuecomment-599135831) if we want.~~ **Edit**: `fastRun.sh` was commited in 8956efc4ae020bb75663be32b73de6fc8d36597b and will be part of **1.8.3**);
3. **often, user reported problems are due to poorly installed environments**. This could help reduce the confusion.

## Possible problems
1. **compatiblity with installing plugins** has to be checked.

   Plugins are internally installed via `npm install`, which would render inconsistent the contents of `node_modules` and `package-lock.json`.
   On the other side, plugins seem to end up in a different directory.

   This really has to be checked.

2. among other things, being more strict means that:

   > **If dependencies** in the package lock **do not match** those in package.json, **npm ci will exit with an error**, instead of updating the package lock.

   This **may confuse our users**, but it's better to have a clear error which a clear solution (== reinstall) than having half-backed installs laying around, or random bug report coming in;

3. something different I may be overlooking

``npm ERR! cipm can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing.``

```
Error: Cannot find module '../../../src/node_modules/wd'

Require stack:

- /home/travis/build/ether/[secure]/tests/frontend/travis/remote_runner.js

    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:982:15)

    at Function.Module._load (internal/modules/cjs/loader.js:864:27)

    at Module.require (internal/modules/cjs/loader.js:1044:19)

    at require (internal/modules/cjs/helpers.js:77:18)

    at Object.<anonymous> (/home/travis/build/ether/[secure]/tests/frontend/travis/remote_runner.js:2:10)

    at Module._compile (internal/modules/cjs/loader.js:1158:30)

    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1178:10)

    at Module.load (internal/modules/cjs/loader.js:1002:32)

    at Function.Module._load (internal/modules/cjs/loader.js:901:14)

    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:74:12) {

  code: 'MODULE_NOT_FOUND',

  requireStack: [

    '/home/travis/build/ether/[secure]/tests/frontend/travis/remote_runner.js'

  ]

}

tests/frontend/travis/runner.sh: 14: kill: No such process

The command "tests/frontend/travis/runner.sh" exited with 1.
cache.2

store build cache

Done. Your build exited with 1.
```
Don't you get the Travis logs?  You need to see them :D

With ``ci`` set in installDeps.sh

## npm cli
```
rm -Rf node_modules/mammoth
rm -Rf src/node_modules
npm install ep_mammoth
bin/run.sh
```
WORKS

## git
```
rm -Rf node_modules/mammoth
rm -Rf node_modules/ep_mammoth
git clone git@github.com:JohnMcLear/ep_mammoth.git
rm -Rf node_modules/mammoth
rm -Rf src/node_modules
bin/run.sh
```
FAILS

```
[2020-04-23 18:37:07.961] [ERROR] console - Failed to load 'ep_mammoth/import.js' for 'ep_mammoth/ep_mammoth/hooks/import': Error: Cannot find module '../../node_modules/mammoth'
Require stack:
- /home/jose/develop/node_modules/ep_mammoth/import.js
- /home/jose/develop/src/static/js/pluginfw/shared.js
- /home/jose/develop/src/static/js/pluginfw/plugins.js
- /home/jose/develop/src/node/server.js
```

Changing from ``ci`` to ``install`` for line 103 in ``installDeps.sh``
```
rm -Rf node_modules/mammoth
rm -Rf node_modules/ep_mammoth
git clone git@github.com:JohnMcLear/ep_mammoth.git
rm -Rf node_modules/mammoth
rm -Rf src/node_modules
bin/run.sh
```

FAILS

```
[2020-04-23 18:37:07.961] [ERROR] console - Failed to load 'ep_mammoth/import.js' for 'ep_mammoth/ep_mammoth/hooks/import': Error: Cannot find module '../../node_modules/mammoth'
Require stack:
- /home/jose/develop/node_modules/ep_mammoth/import.js
- /home/jose/develop/src/static/js/pluginfw/shared.js
- /home/jose/develop/src/static/js/pluginfw/plugins.js
- /home/jose/develop/src/node/server.js

```

## Using WebUI
Installing plugin using the web ui works as expected.

# Conclusion

1. ``git clone`` without doing an ``npm install`` is never a reliable way to install a plugin that has any dependencies.

1. Using ``ci`` in installDeps.sh seems just as reliable as ``install`` as far as plugins go.

# Concerns

None.  IMHO ``ci`` is fine to be merged into 1.8.4

--- a/bin/installDeps.sh
+++ b/bin/installDeps.sh
@@ -100,7 +100,7 @@ echo "Ensure that all dependencies are up to date...  If this is the first time 
   cd node_modules
   [ -e ep_etherpad-lite ] || ln -s ../src ep_etherpad-lite
   cd ep_etherpad-lite
-  npm install --save --loglevel warn
+  npm ci
 ) || {
   rm -rf src/node_modules
   exit 1

--- a/bin/installOnWindows.bat
+++ b/bin/installOnWindows.bat
@@ -14,7 +14,7 @@ cd /D node_modules
 mklink /D "ep_etherpad-lite" "..\src"
 
 cd /D "ep_etherpad-lite"
-cmd /C npm install --loglevel warn || exit /B 1
+cmd /C npm ci || exit /B 1
 
 cd /D "%~dp0\.."
 

